Update data model nav

kdoh opened the pull request:
Changes to the data model to make it easier to view all your segments and metrics across databases or tables.

<img width="1468" alt="Screen Shot 2020-06-09 at 11 50 22 AM" src="https://user-images.githubusercontent.com/5248953/84170295-850f6500-aa47-11ea-846b-64fc02992dd5.png">


Things to un-snarl

- Now that there are specific listing pages for each, I think we should move the segment and metric descriptions into the listing endpoint instead of having to have metadata down on the client to generate them.
- You can now create segments and metrics from the listings but as a result we don't have a way to pre-populate the segment / metric editor with the table you hit `Add segment` from. We'll need to add a data picker to both.


ToDos

- [x] Add metric list view
- [x] Clean up style of database selector on the db view
- [x] Make tab selected state correspond to the url
- [x] List view empty states

kdoh committed:
add basic datamodel app wrapper component

kdoh committed:
rm segments from table view, add new listing page

kdoh committed:
add basic metric list view

kdoh committed:
highlight current tab

kdoh committed:
switch to select for current db

kdoh committed:
remove old metric view on table view

kdoh committed:
lint

robdaemon commented:
@kdoh ping me on Monday and I can help out with the backend portion of this!

paulrosenzweig committed:
Merge branch 'master' into update-datamodel-nav

paulrosenzweig committed:
add dataselector to pick db/table for segments/metrics

paulrosenzweig commented:
I added a data selector on top to populate the metric/segment forms and filter the list. There are still some things I need to fix though (e.g. retiring segments/metrics).

paulrosenzweig committed:
fix archiving and viewing revisions

paulrosenzweig committed:
delete unused files, reorganize things

paulrosenzweig committed:
more reorg

paulrosenzweig committed:
fix cypress tests

paulrosenzweig committed:
remove console.log

paulrosenzweig committed:
use database data selector rather than raw select

paulrosenzweig committed:
test fixes

paulrosenzweig committed:
fix typo in test

Commit:
backend support for updating nav (#12764)

* backend support for updating nav

making functions to move the human-friendly descriptions of filters and
aggregations and such to the server-side

* the basic outline of the descriptor generators is in

* adding segment support

* wiring to frontend

* filling in the description on the forms

* clean namespace forms

* working on tests, appeasing the linters

* testing / fixes

no more ISeq error, but has edge cases where query syntaxes are slightly
different. will resolve via the mbql parsing helpers to avoid (first
(second (second ...))) fun.

* reworked with mbql query helpers

Matches the parts of a query and outputs a data structure that can be
used in the UI

UI wiring is done for display

* remove unused TableMetadata references

paulrosenzweig committed:
prettier

paulrosenzweig committed:
return [] from /metric|segment rather than an empty response

paulrosenzweig committed:
workaround dataselector issue in tests

mazameli commented:
It wasn't clear to me that I needed to select a table on the list view in order to make the "New segment" button clickable. I think this incorrectly conflates filtering the list view with picking the data to base a new segment on. 

![image](https://user-images.githubusercontent.com/2223916/86281896-1aa79b80-bb93-11ea-87e7-3b3c875f2f43.png)

My strong assumption was that I'd be able to pick the table from this first element in the query builder here where it says `Orders`:

![image](https://user-images.githubusercontent.com/2223916/86281993-3d39b480-bb93-11ea-8e7a-fce50f9e4ec2.png)

Going back to the list view, I also strongly assumed that I'd be able to view _all_ of my instance's segments or metrics from these list views without _having_ to filter by a table first; that's one of the benefits of this reorg, as otherwise you have to hunt and peck through your schema to figure out which tables have any segments or metrics. To be explicit, I think that table filtering element should stay; I just don't think that using it should be required.

![image](https://user-images.githubusercontent.com/2223916/86282258-bd601a00-bb93-11ea-9b9d-8c58eb04b100.png)

robdaemon commented:
@mazameli - I had a similar thought to you on that - Maybe clicking "new segment" shows the table chooser?

I guess having a way to filter by table (or sort by table) would be good for customers that have a lot of metrics or segments too?

paulrosenzweig committed:
open segment/metric form without preselected table

paulrosenzweig commented:
@mazameli I made those updates.

> My strong assumption was that I'd be able to pick the table from this first element in the query builder

For metric/segment creation, there's a data selector in that spot in the QB. For existing metrics/segments, it's locked and you can't click it. That should prevent breaking existing queries.

---

> I also strongly assumed that I'd be able to view all of my instance's segments or metrics from these list views without having to filter by a table first

I like this change, but now we probably need a way to deselect a chosen table and return to an unfiltered list. I'm guessing an "x" to the left of the table name is the way to go?

paulrosenzweig committed:
eslint

mazameli commented:
@paulrosenzweig Sweet. The new-segment and new-metric data selection feels exactly right to me now.

> I like this change, but now we probably need a way to deselect a chosen table and return to an unfiltered list. I'm guessing an "x" to the left of the table name is the way to go?

Yeah, ideally I'd like to follow our dashboard filters UI pattern:

![image](https://user-images.githubusercontent.com/2223916/86396780-7176a900-bc57-11ea-9c1c-fcb82accbeb1.png)

![image](https://user-images.githubusercontent.com/2223916/86396998-d7633080-bc57-11ea-8d8f-6bd94c1f9ca9.png)

paulrosenzweig commented:
@mazameli I couldn't get the legend on `<FieldSet>` to work without ugly jumping around as it was added/removed, so I took it out:

![gif](https://user-images.githubusercontent.com/691495/86411667-5c1b7200-bc8b-11ea-9d8f-4e52cd080a5e.gif)

paulrosenzweig committed:
add ability to clear out table filter

mazameli committed:
bold text for filter

paulrosenzweig commented on frontend/src/metabase/query_builder/components/AggregationPopover.jsx:
```diff
@@ -250,14 +250,15 @@ export default class AggregationPopover extends Component {
+            : // GA metrics are more like columns, so they should be displayed even when showMetrics is false
+              metric.googleAnalyics,
+        )
+      : [];
```
@tlrobinson in lieu of a larger review, what do you think about putting this here? Previously we were creating a `new Metadata` that filtered out non-Google Analytics metrics. That felt ugly and I figured the PartialQueryBuilder should use the full Metadata and filter it when displaying. Is there a better spot for this logic?

Definitely feel free to comment on whatever else here if you want, but there's a big diff and most of this PR is moving things around and small refactors.

tlrobinson replied:
Yeah I think this is better than a custom Metadata

paulrosenzweig committed:
Merge branch 'master' into update-datamodel-nav

paulrosenzweig committed:
add back empty list text

paulrosenzweig reviewed (approved):
This is somewhat a self-approval, but three of us have worked on this together. When the tests pass, I'll merge so we can get the 36 RC out.
